Action History Plugin

Vijar opened the pull request:
@mridgway @redonkulus @lingyan 

this implements the same thing as my other PR but in a much more uglier way by wrapping it in a plugin.

lingyan commented:
Nice!  I totally like the way this is going 💃

lingyan commented on packages/fluxible-plugin-perf/lib/perf-plugin.js:
```diff
@@ -0,0 +1,115 @@
+            var actionHistory = [];
+            console.log('perf enabled', enableDebug);
+            console.log(context);
+            context.getActionHistory = function () {
```
I suggest we namespace the function with `context.perfTools.` (or `context.devTools.`)

mridgway commented:
Looks great. I think it would be great to add a React component with the action graph as well.

I think we can use this package as a place for future dev tools as well. So maybe `fluxible-plugin-devtools` is a better name.

yahoocla commented:
CLA is valid!

kaesonho commented:
:+1:

mridgway commented on packages/fluxible-plugin-perf/index.js:
```diff
@@ -0,0 +1,6 @@
+ * Copyright 2014, Yahoo! Inc.
+ * Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
+ */
+console.log('perf-plugin');
```
Remove.

mridgway commented on packages/fluxible-plugin-perf/lib/perf-plugin.js:
```diff
@@ -0,0 +1,115 @@
+        /**
+         * @property {String} name Name of the plugin
+         */
+        name: 'fPerfPlugin',
```
Is this a typo?

mridgway commented on packages/fluxible-plugin-perf/lib/perf-plugin.js:
```diff
@@ -0,0 +1,115 @@
+            var enableDebug = typeof contextOptions.debug !== 'undefined' ? contextOptions.debug : false;
+            var actionHistory = [];
+            console.log('perf enabled', enableDebug);
+            console.log(context);
```
Remove.

mridgway commented on packages/fluxible-plugin-perf/lib/perf-plugin.js:
```diff
@@ -0,0 +1,115 @@
+                 * Adds the service CRUD and getServiceMeta methods to the action context
+                 * @param actionContext
+                 */
+                plugActionContext: function plugActionContext(actionContext) {
```
Remove.

mridgway commented on packages/fluxible-plugin-devtools/index.js:
```diff
@@ -0,0 +1,8 @@
+ * Copyright 2016, Yahoo! Inc.
+ * Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
+ */
+console.log('perf-plugin');
```
Remove.

mridgway commented on packages/fluxible-plugin-devtools/.gitignore:
```diff
@@ -0,0 +1 @@
+dest/
```
Usually the folder is called `dist` for compiled code.

mridgway commented:
Is d3 a dependency? Is it just expected to be a global?

mridgway commented on gulpfile.babel.js:
```diff
@@ -51,6 +52,26 @@ gulp.task('install', () => {
+        throw new Error('Usage: gulp dist -p <package>');
+    }
+    var src = packages[packageName] + '/src/**/*';
+    var dest = path.resolve(packages[packageName], 'dest')
```
`dist` for the folder name.

mridgway commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+dest
```
`dist`

mridgway commented on packages/fluxible-plugin-devtools/src/lib/devtools-plugin.js:
```diff
@@ -0,0 +1,141 @@
+
+/**
+ * Creates a new devtools plugin instance
+ * @returns {PerfPlugin}
```
`PerfPlugin` -> `DevToolsPlugin`

mridgway commented on package.json:
```diff
@@ -25,6 +25,7 @@
     "babel-plugin-transform-object-rest-spread": "^6.6.4",
     "babel-preset-es2015": "^6.6.0",
     "babel-preset-react": "^6.5.0",
+    "babel-preset-stage-1": "^6.5.0",
```
What are you using from stage-1 that wasn't included individually? I'd prefer to do individual features so we can keep track of what is used I think.

Vijar commented:
@mridgway we load d3 from cdn for now. I didn't want to create an explicit dependency because I don't want d3 to be bundled in webpack bundles.

Vijar commented:
As I was adding tests, I realized that my implementation was actually incorrect. While the start times were correct, I was attributing action end times to the parent action instead of the current executing action.

I fixed it, but the solution is more complicated than before. I opted to still maintain this code in a plugin and added a feature into fluxible to allow plugging the arguments to the internal `callAction` method. I'm open to suggestions on how to better accomplish this though.

Here is the before:
<img width="759" alt="screen shot 2016-04-10 at 12 43 23 am" src="https://cloud.githubusercontent.com/assets/759766/14408762/f2f461c0-feb5-11e5-8a2e-17be3c3729ab.png">
And after my fix:
<img width="755" alt="screen shot 2016-04-10 at 12 44 13 am" src="https://cloud.githubusercontent.com/assets/759766/14408764/fc2a787e-feb5-11e5-9c4a-cd240637d040.png">

You can see how the leaf nodes had no width before but now they're correctly timed.

mridgway commented on packages/fluxible-plugin-devtools/src/lib/devtools-plugin.js:
```diff
@@ -0,0 +1,169 @@
+                    }
+
+                    const action = options.action;
+                    const actionReference = action.__actionReference
```
I don't think it's safe to attach on to the `action` in this way in case there are multiple executions of the same action at the same time.

Vijar replied:
@mridgway you're right, that was unnecessary.

redonkulus commented:
@Vijar would love to see this integrated into the site and /docs/ page for the site as well.

redonkulus commented on packages/fluxible-plugin-devtools/src/components/ActionTree.jsx:
```diff
@@ -0,0 +1,271 @@
+            return;
+        }
+        var relativeWidth = this.props.relativeWidth;
+        console.log('relativeWidth', relativeWidth);
```
remove

mridgway commented on packages/fluxible/lib/FluxibleContext.js:
```diff
@@ -113,6 +113,21 @@ function executeActionProxy(context, actionContext, action, payload, done) {
                 '` will only start after `' + currentActionDisplayName + '` is complete.');
         }
     }
+    context._plugins.forEach(function pluginsEach(plugin) {
```
Can we accumulate the `plugExecuteAction` plugins upon context creation so that we don't have to iterate over all of the plugins for every `executeAction`. In most cases, this will be empty even though there will be multiple plugins installed (fetchr, perf, i13n, etc.) and I don't want to add to the overhead of each action since it is a hot path.

redonkulus commented on packages/fluxible-plugin-devtools/src/lib/devtools-plugin.js:
```diff
@@ -0,0 +1,169 @@
+             */
+            function overrideCtx() {
+                debug('devtools plugin %s', enableDebug ? 'enabled' : 'disabled')
+                if(!enableDebug) {
```
add space after if

mridgway commented on packages/fluxible-plugin-devtools/src/components/ActionTree.jsx:
```diff
@@ -0,0 +1,271 @@
+            .attr("transform", `translate(${margin.left},${margin.top})`);
+
+        function update(source) {
+            console.log('updating');
```
Remove.

mridgway commented on packages/fluxible-plugin-devtools/src/components/ActionTree.jsx:
```diff
@@ -0,0 +1,271 @@
+
+            // Compute the flattened node list. TODO use d3.layout.hierarchy.
+            var nodes = tree.nodes(source);
+            console.log('nodes', source, nodes);
```
Remove.

mridgway commented on packages/fluxible-plugin-devtools/src/components/ActionTree.jsx:
```diff
@@ -0,0 +1,271 @@
+                const durPercent = n.parent ? (n.duration / source.duration) : 1;
+                const width = Math.round(durPercent * barWidth) || 1; // minimum 1 pixel width
+                n.width = relativeWidth ? width : barWidth;
+                console.log(i, n.name, n.x, i*barHeight);
```
Remove.

mridgway commented on packages/fluxible-plugin-devtools/tests/unit/lib/devtools-plugin.js:
```diff
@@ -0,0 +1,250 @@
+                };
+                var heirarchy = context.getComponentContext().devtools.getActionHistory()[0];
+                function compareHeirarchy(expected, actual) {
+                    console.log('comparing', expected, actual);
```
Remove.

redonkulus commented on packages/fluxible-plugin-devtools/src/lib/devtools-plugin.js:
```diff
@@ -0,0 +1,169 @@
+             * @param {Object} obj arbitrary input object
+             * @return {void}
+             */
+            function provideDevTools (obj) {
```
does this function need to be created in plugContext or can it be moved outside the exports?

Vijar replied:
Needs to be inside. This function is a closure with access to the `actionHistory` var. Otherwise, we would just repeat the contents of the function multiple times.

redonkulus commented on packages/fluxible-plugin-devtools/src/lib/devtools-plugin.js:
```diff
@@ -0,0 +1,169 @@
+             * @method overridectx
+             * @return {void}
+             */
+            function overrideCtx() {
```
does this function need to be created in plugContext or can it be moved outside the exports?

Vijar replied:
Needs to be inside. This function is a closure with access to the `context` var when called from rehydrate.

Vijar commented:
@redonkulus added docs and stuff. 

We can run this on fluxible.io using the query param `?debug=1`. Here is what it looks like:
<img width="1226" alt="screen shot 2016-04-12 at 4 48 30 pm" src="https://cloud.githubusercontent.com/assets/759766/14479394/c7db5d94-00d2-11e6-80e7-05337e540d4e.png">

mridgway commented:
Looks good to me.

redonkulus commented on site/components/Debug.js:
```diff
@@ -0,0 +1,65 @@
+
+    componentDidMount() {
+        const query = this.context.query;
+        console.log('mounted', query.debug);
```
remove

redonkulus commented on site/components/Debug.js:
```diff
@@ -0,0 +1,65 @@
+
+    render() {
+        if (!this.state.shouldRender) {
+            console.log('should not rener');
```
remove

redonkulus commented:
Minor comments that can be fixed later 👍
